File: algos/src/msg.rs
use std::io::{Cursor, Read};

use byteorder::{LittleEndian, ReadBytesExt};
use serde::{Deserialize, Serialize};

use crate::{doc::Doc, pid::Pid};


#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum PeerMessage {
    Greet,
    Insert { site: u8, pid: Pid, c: char },
    Delete { site: u8, pid: Pid },
    NewSession { site: u8, doc: Doc },
}

impl PeerMessage {
    pub fn serialize(&self) -> Vec<u8> {
        match self {
            PeerMessage::Greet => vec![0u8],
            PeerMessage::NewSession { site, doc } => {
                // put header i.e. PeerMessage enum it is
                let mut buf = vec![1u8];
                // put site_id
                buf.push(*site);
                // put numberofatoms
                buf.extend((doc.len() as u64).to_le_bytes());
                doc.write_bytes_tobuf(&mut buf);
                buf
            }
            PeerMessage::Insert { site, pid, c } => {
                let mut buf = vec![2u8];
                buf.push(*site);
                let mut cbuf = [0u8; 4];
                let encoded = c.encode_utf8(&mut cbuf);
                // put atom's data length
                buf.push(encoded.len() as u8);
                buf.extend(encoded.as_bytes());
                buf.push(pid.depth() as u8);
                pid.write_bytes(&mut buf);
                buf
            }
            PeerMessage::Delete { site, pid } => {
                let mut buf = vec![3u8];
                buf.push(*site);
                buf.push(pid.depth() as u8);
                pid.write_bytes(&mut buf);
                buf
            }
        }
    }
    pub fn deserialize(buf: &[u8]) -> PeerMessage {
        let mut cur = Cursor::new(buf);
        match cur.read_u8().unwrap() {
            0u8 => PeerMessage::Greet,
            1u8 => {
                let site = cur.read_u8().unwrap();
                let number_of_atoms = cur.read_u64::<LittleEndian>().unwrap() as usize;
                PeerMessage::NewSession {
                    site: site,
                    doc: Doc::from_reader(&mut cur, number_of_atoms),
                }
            }
            2u8 => {
                let site = cur.read_u8().unwrap();
                let data_len = cur.read_u8().unwrap() as usize;
                let mut bytes = [0u8; 4];
                cur.read_exact(&mut bytes[..data_len]).unwrap();
                let data = std::str::from_utf8(&bytes[..data_len])
                    .unwrap()
                    .chars()
                    .next()
                    .unwrap();
                let pid_depth = cur.read_u8().unwrap();
                let pid = Pid::from_reader(&mut cur, pid_depth as usize);
                PeerMessage::Insert { site: site, pid: pid, c: data }
            }
            3u8 => {
                let site = cur.read_u8().unwrap();
                let pid_depth = cur.read_u8().unwrap();
                let pid = Pid::from_reader(&mut cur, pid_depth as usize);
                PeerMessage::Delete { site: site, pid: pid }
            }
            _ => panic!(),
        }
    }
}

File: kotlingoot/app/src/main/kotlin/org/example/Msg.kt
package org.example
import Doc
import Pid
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.DataInputStream
import java.io.DataOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

sealed class PeerMessage {
    object Greet : PeerMessage()

    data class Insert(val site: UByte, val pid: Pid, val c: Char) : PeerMessage()
    data class Delete(val site: UByte, val pid: Pid) : PeerMessage()
    data class NewSession(val site: UByte, val doc: Doc) : PeerMessage()

    fun serialize(): ByteArray {
        return when (this) {
            is Greet -> byteArrayOf(0u.toByte())

            is NewSession -> {
                val buf = ByteArrayOutputStream()
                buf.write(1)
                buf.write(site.toInt())
                val lenBytes = ByteBuffer.allocate(8)
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .putLong(doc.len().toLong())
                    .array()
                buf.write(lenBytes)
                doc.writeBytes(buf)
                buf.toByteArray()
            }

            is Insert -> {
                val buf = ByteArrayOutputStream()
                buf.write(2)
                buf.write(site.toInt())
                val encoded = c.toString().toByteArray(Charsets.UTF_8)
                buf.write(encoded.size)
                buf.write(encoded)
                buf.write(pid.depth().toInt())
                pid.writeBytes(buf)
                buf.toByteArray()
            }

            is Delete -> {
                val buf = ByteArrayOutputStream()
                buf.write(3)
                buf.write(site.toInt())
                buf.write(pid.depth().toInt())
                pid.writeBytes(buf)
                buf.toByteArray()
            }
        }
    }

    companion object {
        fun deserialize(buf: ByteArray): PeerMessage {
            val input = DataInputStream(ByteArrayInputStream(buf))
            return when (val tag = input.readUnsignedByte()) {
                0 -> Greet

                1 -> {
                    val site = input.readUnsignedByte().toUByte()
                    val numberOfAtoms = java.lang.Long.reverseBytes(input.readLong()).toInt()
                    val doc = Doc.fromReader(input, numberOfAtoms)
                    NewSession(site, doc)
                }

                2 -> {
                    val site = input.readUnsignedByte().toUByte()
                    val dataLen = input.readUnsignedByte()
                    val bytes = ByteArray(4)
                    input.read(bytes, 0, dataLen)
                    val c = bytes.decodeToString(0, dataLen).first()
                    val pidDepth = input.readUnsignedByte()
                    val pid = Pid.fromReader(input, pidDepth)
                    Insert(site, pid, c)
                }

                3 -> {
                    val site = input.readUnsignedByte().toUByte()
                    val pidDepth = input.readUnsignedByte()
                    val pid = Pid.fromReader(input, pidDepth)
                    Delete(site, pid)
                }

                else -> throw IllegalArgumentException("Unknown message tag: $tag")
            }
        }
    }
}
